examples/cgsolve.cu

```diff
@@ -83,6 +88,10 @@ int main([[maybe_unused]] int argc, [[maybe_unused]] char **argv)
   (maxn = matx::max(sqrt(norm))).run(exec);
 
   exec.sync();
+#if 1
+  ctx.finalize();
```

Comment:
what is `finalize` used for vs `sync`? Could you hide the context in the executor so the user doesn't need it, and calling `exec.sync()` calls `finalize()`?

Reply:
finalize terminates everything in the stf context, it waits for asynchronous tasks, deletes internal resources etc...  you can only do it once, sync is more equivalent to a ctx.task_fence() which is a non blocking fence (it returns a CUDA stream, and waiting for that stream means everything was done).

I'd like to move finalize to the dtor of the executor, but there are some caveats if you define the executor as a static variable, is this allowed ? The caveat might be some inappropriate unload ordering of CUDA and STF libraries as usual ...

Reply:
Sounds good. I think the destructor is the right place. but does sync() work as expected?

Reply:
@sidelnik is it doing a task fence with a stream sync ?

Reply:
@caugonnet , sync() should be calling ctx.task_fence() now. I agree, I think we should place the ctx.finalize() inside the stf executor dtor